src/Avalonia.Base/Data/Core/BindingExpression.cs

```diff
@@ -1,389 +1,549 @@
-using System;
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
 using System.Diagnostics.CodeAnalysis;
 using System.Globalization;
-using Avalonia.Reactive;
+using System.Linq.Expressions;
+using System.Text;
 using Avalonia.Data.Converters;
+using Avalonia.Data.Core.ExpressionNodes;
+using Avalonia.Data.Core.Parsers;
+using Avalonia.Input;
+using Avalonia.Interactivity;
 using Avalonia.Logging;
 using Avalonia.Utilities;
 
-namespace Avalonia.Data.Core
+namespace Avalonia.Data.Core;
+
+/// <summary>
+/// A binding expression which accepts and produces (possibly boxed) object values.
+/// </summary>
+/// <remarks>
+/// A <see cref="BindingExpression"/> represents a untyped binding which has been
+/// instantiated on an object.
+/// </remarks>
+internal partial class BindingExpression : UntypedBindingExpressionBase, IDescription, IDisposable
 {
+    private readonly WeakReference<object?>? _source;
```

Comment:
Very happy to see this PR!

```suggestion
    ~BindingExpression()
    {
        _source.Free(); // field must not be readonly: this call mutates its value
    }
    
    private GCHandle _source;
```

For private fields, it's reasonable to use `GCHandle` directly and replace the `WeakReference` finalizer with our own. This eliminates one object allocation per binding expression.

Reply:
Probably worth doing in a separate PR so it can be reverted easily if it causes a problem.